src/StreamJsonRpc/Reflection/MessageFormatterProgressTracker.cs

```diff
@@ -0,0 +1,168 @@
+
+namespace StreamJsonRpc.Reflection
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+    using Microsoft;
+    using Microsoft.VisualStudio.Threading;
+    using Newtonsoft.Json.Linq;
+
+    /// <summary>
+    /// Class containing useful methods to help message formatters implement support for <see cref="IProgress{T}"/>.
+    /// </summary>
+    public class MessageFormatterProgressTracker
+    {
+        /// <summary>
+        /// Dictionary used to map progress id token to its corresponding <see cref="ProgressParamInformation" /> instance containing the progress object and the necessary fields to report the results.
+        /// </summary>
+        public readonly Dictionary<long, ProgressParamInformation> ProgressMap = new Dictionary<long, ProgressParamInformation>();
+
+        /// <summary>
+        /// Special method name for progress notification.
+        /// </summary>
+        internal const string ProgressRequestSpecialMethod = "$/progress";
+
+        /// <summary>
+        /// Object used to lock the access to <see cref="RequestProgressMap"/> and <see cref="ProgressMap"/>.
+        /// </summary>
+        internal readonly object ProgressLock = new object();
+
+        /// <summary>
+        /// Dictionary used to map the outbound request id to their progress id token so that the progress objects are cleaned after getting the final response.
+        /// </summary>
+        internal readonly Dictionary<long, long> RequestProgressMap = new Dictionary<long, long>();
+
+        /// <summary>
+        /// Gets or sets the the next id value to assign as token for the progress objects.
+        /// </summary>
+        internal long NextProgressId { get; set; }
+
+        /// <summary>
+        /// Gets or Sets the id of the request currently being serialized so the converter can use it to create the request-progress map.
+        /// </summary>
+        internal long? RequestIdBeingSerialized { get; set; }
+
+        /// <summary>
+        /// Converts given <see cref="Type"/> to its <see cref="IProgress{T}"/> type.
+        /// </summary>
+        /// <param name="objectType">The type which may implement <see cref="IProgress{T}"/>.</param>
+        /// <returns>The <see cref="IProgress{T}"/> from given <see cref="Type"/> object, or <c>null</c>  if no such interface was found in the given <paramref name="objectType" />.</returns>
+        public static Type FindIProgressOfT(Type objectType)
+        {
+            Requires.NotNull(objectType, nameof(objectType));
+
+            if (objectType.IsConstructedGenericType && objectType.GetGenericTypeDefinition().Equals(typeof(IProgress<>)))
+            {
+                return objectType;
+            }
+
+            return objectType.GetTypeInfo().GetInterfaces().FirstOrDefault(i => i.IsConstructedGenericType && i.GetGenericTypeDefinition() == typeof(IProgress<>));
+        }
```

Comment:
this is a relatively expensive set of code to run for every single value being deserialized (at least for the JSON formatter, the way the JsonConverter is used). I think we should probably build up a static `Dictionary<Type, Type>` and record/recall every calculation we do here (with thread-safety in mind). That should help prevent perf regressions from this change, which we want to avoid since it will impact all messages, even if they don't use `Progress<T>`.